Editor: if textbox is focused, edit menu affects it

ericoporto opened the pull request:
additional for #2231 

There are still small details there, but in this way, using an action in the edit menu won't affect the scintilla editor if a textbox has focus.

Checking for focus in scintilla instead won't work, so this is why it's done in such way.

ivan-mogilko commented:
But this is not enough for properly applying the action over the textbox. Currently for scintilla it also checks for availability. For example: whether it is possible to cut or copy, whether it is possible to undo or redo, and enables and disables the commands accordingly. With this change the commands will act inconsistently.

ericoporto commented:
> applied on controls that may not belong to the panel

So this can't happen because if a different pane is focused the AGS Editor will readjust the menus to that pane, so while that focus query can get focus from any control in the editor, it's only called when something in the pane is already in focus.

> Make the commands enabled only when Scintilla is active.

I made an attempt at something like this, if scintilla is gets focus, the edit menu gets updated.

> If you like them to apply to the Dialog options too, then this should be handled inside DialogEditor (as it's the class that knows that it has other input fields), and make sure that it affects only its children

Took a stab at that.

> For the real proper behavior it would also need to check which control is under focus right now, and change the command states according to the active control and its state.

OK, I thought that was a bit too much, and instead just enable everything if in a textbox.

ivan-mogilko commented:
> > applied on controls that may not belong to the panel

> So this can't happen because if a different pane is focused the AGS Editor will readjust the menus to that pane, so while that focus query can get focus from any control in the editor, it's only called when something in the pane is already in focus.

This can happen (and happens!) if a text box is not on a ContentDocument control. For instance, if it's a Properties Grid.

ivan-mogilko commented:
I still do not understand why the textbox handling has to be inside ScriptEditorBase class: that is a parent class for a script editor, which is meant to handle only Scintilla control. It does not know that there may be any textboxes.
If anything, then DialogEditor should be handling its own textboxes.

But, the real problem is that this PRs goal is not well defined imo, and it's not working consistently.
First of all, why there's a need to make these Edit menu options work on the option text boxes at all?
Secondly, what about the other options, below the Copy/Paste etc? They are still active in this menu when text box is focused, but they do not work when you click on them. I'm speaking about "Find", "Goto Line", "Show autocomplete", and others.

I suppose it would be simpler to disable this menu completely when the focus leaves scintilla.

ivan-mogilko commented:
I still think that the textbox handling should not be done inside ScriptEditorBase class: that is a parent class for a script editor, which is meant to handle only Scintilla control. It does not know that there may be any textboxes or anything else.
If anything, then DialogEditor should be handling its own textboxes.

It is possible to override OnCommandClick() in DialogEditor, and intercept some commands. Possibly these command constants in ScriptEditorBase have to be turned from `private` into `protected` to let the derived classes know about those.
Then you could handle these differently in DialogEditor.

UpdateUICommands() may also be overridden as necessary, instead of passing a slightly strange "force" parameter into the base class.

Besides the above, there's an issue with the rest of the menu: commands such as "Find", "Goto Line", "Show autocomplete", and others. They are kept enabled while the focus is in the textbox, but do not do anything.
IMO some of those, like "Goto Line" and "Find", may keep working even if focus is in the textbox, since it's obvious they are about the script.
If OnCommandClick() UpdateUICommands() were handled in DialogEditor class, then it could accurately determine which commands to disable and which to keep, which should have special meaning for textboxes, and which to delegate to the base class.

ericoporto commented:
I moved the textbox things out of the ScriptEditorBase and into the DialogEditor and added additional code in the dialog editor to make sure the Edit menu updates correctly when the cursor is in either the scintilla or in a textbox. I am not dealing with the other things it could be because there isn't really a good way to do so. Using something that is not cut/copy/paste/undo/redo will force focus into Scintilla, even if it's not in scintilla before.

ivan-mogilko commented:
I think this might be fine, but could you add a UpdateUICommands after DialogEditor does OnCommandClick, similar to how it's done in the base class? This lets to enable Undo command, otherwise it's stuck in disabled state unless you switch to another text box and return back.

ericoporto commented:
OK, added also an extra KeyUp event there because I noticed it picked almost everything except selecting text with the keyboard arrows + shift! Now it's really checking all possible situations!

ivan-mogilko commented:
Unfortunately I missed this, but `force` argument in UpdateUICommands has to be redone somehow. This is a virtual method, which means that it may be overidden in subclasses, but "force" only has a meaning for limited number of commands in the base class.
